### PR TITLE
Optimise Library page load time when channels have large Thumbnail

### LINKED ISSUE #12502: Channel thumbnails can cause extreme slowdown of channel API call
## Observed behavior
When loading the available channels in Kolibri, the inclusion of the thumbnails in the API response can cause it to be very large, blocking page rendering. Without the thumbnails, a 1.6MB API payload can be reduced to less than 6KB. This is a significantly larger saving than just gzipping the payload, which reduces by about 25%.

## Expected behavior
Instead of returning the thumbnails as base64 encoded images, the internal channelmetadata API endpoint should return a URL for a new view that reads the contents of the the thumbnail column of the channel metadata table, and returns the content as a file response.

This will allow the thumbnails to be loaded in a non-blocking way allowing significantly faster load times wherever channels are loaded.

N.B. For now this should not affect the public channel metadata endpoints, as older Kolibris will still expect the base64 encoded version. A follow up issue will be filed to better handle this for the public APIs.

## User-facing consequences
Getting rid of timeouts on the library page when users have lots of channels with large thumbnails.

## Context
Kolibri version 0.16.2, with many channels with large thumbnails installed.

### PR DESCRIPTION
**Pull request by thesujai**
## Summary

As mentioned in the issue this PR does the following changes:

1. Defines a new API endpoint to get the `channel_thumbnail` as a img file
2. Changes the Current internal `ChannelMetaData` endpoint to return a URL to thumbnail instead directly returning a base64 encoded URL. This will cause async loading for thumbnails in the frontend

**ScreenCast**
[Screencast from 02-08-24 01:44:04 PM IST.webm](https://github.com/user-attachments/assets/6eadd536-0426-4015-a112-4675cb212eb1)

## References

Fixes #12502 

## Reviewer guidance

Open the library page and observe the channels load immediately, but the corresponding thumbnails loads asynchronously 

----

## Testing checklist

- [x] Contributor has fully tested the PR manually
- [x] If there are any front-end changes, before/after screenshots are included
- [ ] Critical user journeys are covered by Gherkin stories
- [x] Critical and brittle code paths are covered by unit tests

## PR process

- [x] PR has the correct target branch and milestone
- [ ] PR has 'needs review' or 'work-in-progress' label
- [ ] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

### CODE REVIEW

File: kolibri/core/content/api.py
Comment (thesujai) [inline kolibri/core/content/api.py:298]: For writing unit test for this is there any existing `setUp` method that will set up a channel metadata model?

File: kolibri/plugins/learn/assets/src/views/ChannelThumbnail.vue
Comment (thesujai) [inline kolibri/plugins/learn/assets/src/views/ChannelThumbnail.vue]: I think we can do a high-level fix here, where we just modify the `channels` on the library page.
Now it is affecting all the cards on the library page. I don't know where to make changes for that.

File: kolibri/core/content/api.py
Comment (rtibbles) [inline kolibri/core/content/api.py]: I think this might be doable using the `field_map` class property which allows setting to a callable, that would allow the rest of the code not to be copy pasted here.

File: kolibri/core/content/api.py
Comment (rtibbles) [inline kolibri/core/content/api.py:405]: So you would add here:
```
field_map = {
    "thumbnail": _create_channel_thumbnail_url,
}

field_map.update(BaseChannelMetadataMixin.field_map)
```

Then in the module scope you can define this function:

```
def _create_channel_thumbnail_url(item):
    return reverse("kolibri:core:channel-thumbnail", args=[item["id"]])
```

File: kolibri/plugins/learn/assets/src/views/ChannelThumbnail.vue
Comment (rtibbles) [inline kolibri/plugins/learn/assets/src/views/ChannelThumbnail.vue]: If the ChannelThumbnail endpoint is working as intended, it should be returning a valid file if fetched - so you should just be able to use the URL directly as a file URL.

If this is not working, we need to make some tweaks, but when it is working correctly, this should require no frontend changes at all.

File: kolibri/core/content/api.py
Comment (rtibbles) [reply kolibri/core/content/api.py:298]: If you look in the `test_content_app.py` file in the `content` app, you should see some examples.

File: kolibri/core/content/api.py
Comment (rtibbles) [inline kolibri/core/content/api.py]: I am not confident this will be sufficient - the file is base64 encoded, which has a prefix with the mimetype. I think it is likely you would need to use the base64 decode method before serving this as the contents of a file: https://docs.python.org/3/library/base64.html#base64.b64decode

**Review by rtibbles — DISMISSED** [review]
This should work without any frontend changes, so let's aim for that.

Changes to the channel metadata endpoint can be simplified, and avoid code duplication.

I suspect that the reason the frontend changes were needed is because the backend endpoint is not returning something that the browser interprets as an image file.

**Comment by thesujai** [comment]
@rtibbles can you take another look?

File: kolibri/core/content/api.py
Comment (thesujai) [reply kolibri/core/content/api.py:405]: Yes this works, can you share django/python docs covering `field_map`?

Comment (rtibbles) [reply kolibri/core/content/api.py:405]: This is not a Django thing - so the documentation for it is in this code comment: https://github.com/learningequality/kolibri/blob/develop/kolibri/core/api.py#L139

File: kolibri/core/content/test/test_content_app.py
Comment (rtibbles) [inline kolibri/core/content/test/test_content_app.py]: Small typo here! `ChannelThumbnailViewTestCase`

File: kolibri/core/content/api.py
Comment (rtibbles) [inline kolibri/core/content/api.py]: Instead of doing a falsey check on line 301, you could instead just wrap this call in try/except for a `ValueError`:
```
try:
    mimetype, b_64_thumbnail = channel.thumbnail.split(",", 1)
except ValueError:
    raise Http404("No thumbnail available")
```

This would have the slight advantage of catching malformed as well as empty thumbnails.

File: kolibri/core/content/api.py
Comment (rtibbles) [inline kolibri/core/content/api.py]: It looks like we are still loading the `thumbnail` data from the database - perhaps we could also do a falsey check here?

```
return reverse("kolibri:core:channel-thumbnail", args=[item["id"]]) if item["thumbnail"] else ""
```

That would mean that we don't set a URL in the case that the thumbnail is empty, and Kolibri can display the default channel icon instead of the thumbnail (whereas if the thumbnail URL 404s, we will see a broken image display in the browser).

**Review by rtibbles — COMMENTED** [review]
A couple of considerations, and a typo.

Have not yet manually tested!

**Comment by thesujai** [comment]
I made the changes.
You can do the manual testing.
Now I can observe that when the library page loads, individual HTTP requests are made to get the thumbnail. 
The `Other libraries` section which I suppose makes requests to the public endpoint is not affected by this change, as I see a base64 encoded thumbnail returned in the browser network tab. That is the desired behavior for now.

**Review by rtibbles — COMMENTED** [review]
Looks good, after manual QA, this will be good to merge (after we've released 0.17.0).

File: kolibri/core/content/api.py
Comment (rtibbles) [inline kolibri/core/content/api.py]: In manual testing, I noticed that this is giving a mimetype of `data:image/png;base64` for example.

Adding:
```
# Split the mimetype to remove data: prefix and ;base64 suffix.
mimetype = mimetype.split(":", 1)[1].split(";", 1)[0]
```
as another line within the try except block, strips it down to just the mimetype.

It doesn't seem to make any difference to rendering, but seems neater to have the mimetype reported precisely.

Comment (thesujai) [reply kolibri/core/content/api.py]: makes sense, updated!

**Comment by radinamatic** [comment]
cc @pcenov

**Comment by pcenov** [comment]
@radinamatic - no issues observed while manually testing, implemented as specified. Good to go!

**Review by rtibbles — APPROVED** [review]
Code changes look good, manual QA checks out, this is ready to go!

**Pull request merged**
